backupccl: backup/restore with user-defined functions

chengxiong-ruan opened the pull request:
Fixes #84087
With user-defined functions introduced, backup/restore needs to work with the new function descriptors and schema descriptors containing function signatures. This commit adds logic into current backup/restore infrastructure to make sure function descriptors are properly backed up and restored.

Release note (enterprise change): backup/restore now can backup and restore user-defined function descriptors at database and cluster level.
Release justification: necessar fix to backup/restore to make sure it works with user-defined functions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87561)
<!-- Reviewable:end -->

postamar commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10303,3 +10305,293 @@ func TestBackupDBWithViewOnAdjacentDBRange(t *testing.T) {
+	require.Equal(t, "1", rows[0][0])
+	require.NoError(t, err)
+
+}
```
Is there any reason why these tests can't be implemented as data-driven tests? See `TestDataDriven`. This would be much more preferable. Also, I wouldn't count too much on individual ID values being stable over time, but that's a secondary concern.

chengxiong-ruan replied:
I was a fool and didn't notice that's a thing for backup/restore :( Will move these tests there and add more tests.

postamar replied:
Ignorance is not foolishness.

postamar commented on pkg/ccl/backupccl/backupresolver/targets.go:
```diff
@@ -590,8 +616,18 @@ func DescriptorsMatchingTargets(
 				}
+				// TODO(chengxiong): get all the user-defined functions used by this
+			// table. This is needed when we start supporting udf references from
+			// other objects.
```
nit: whitespace

postamar reviewed (commented):
This all seems reasonable but I'd be a lot more confident about this PR if it had a lot more tests. I would like to see functions featured in many of the existing data-driven test definitions in `pkg/ccl/backupccl/testdata/backup-restore`.

chengxiong-ruan committed:
sql: check type dependent descriptor type when dropping a type

Previously we assume only tables/views can depend on a type.
So when we check dependencies when dropping a type, we assume
all dependent IDs are table IDs. This is not true with UDF
introduced. We fixed most of similar problem for all other
objects, but forgot about UDT :(

Release note: None.
Release justification: low risk bug fix

chengxiong-ruan committed:
backupccl: backup/restore with user-defined functions

Fixes #84087
With user-defined functions introduced, backup/restore needs
to work with the new function descriptors and schema descriptors
containing function signatures. This commit adds logic into
current backup/restore infrastructure to make sure function
descriptors are properly backed up and restored.

Release note (enterprise change): backup/restore now can backup
and restore user-defined function descriptors at database and
cluster level.
Release justification: necessar fix to backup/restore to make
sure it works with user-defined functions.

chengxiong-ruan commented:
@postamar thanks for the suggestion of using data driven test here, it helped me found 2 small bugs! I added data driven tests for both db backup/restore and cluster backup/restore. Instead of checking IDs, dependency behaviors on DROP/RENAME are tested. I found the existing data driven tests are mostly orthogonal so I didn't touch them, let me know if you have different opinion. I actually like to keep the two unit tests from earlier commit. I understand it'd be annoying that IDs are not stable when, for example, new system tables are added or other objects are added to the test. But I think it's a good idea to see the IDs actually match as an extra line of guard (since we're testing IDs rewritten).

postamar commented:
Thanks, this looks good as far as I can tell. Perhaps add a test case to `TestCleanupDoesNotDeleteParentsWithChildObjects`? Just to check the behaviour is as expected when a restore fails. 

This release-blocker needs review by the Disaster Recovery team, please. CC @stevendanna @dt

chengxiong-ruan committed:
backupccl: drop function if restore failed

This commit adds logic to cleaning logic on restore failures,
so that function descriptors are dropped if restore failed.

Release note: None.
Release justification: necessary bug fix to make sure backup
and restore work with udf.

chengxiong-ruan commented:
> Thanks, this looks good as far as I can tell. Perhaps add a test case to `TestCleanupDoesNotDeleteParentsWithChildObjects`? Just to check the behaviour is as expected when a restore fails.

That's a good catch! udf was not dropped for failed restore. Fixed that and added test cases for it.

stevendanna reviewed (commented):
Thanks for thinking about backup & restore.

I think we want to add a check for USAGE during backup here: https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/backupccl/backup_planning.go#L466

Overall this looks reasonable to me.  I suppose that UDFs may references tables that aren't being backed up, but that matches how we handle foreign keys and such currently.

chengxiong-ruan commented on pkg/ccl/backupccl/restore_planning.go:
```diff
@@ -1745,10 +1782,18 @@ func doRestorePlan(
+	if newDBName != "" {
+		overrideDBName = newDBName
+	}
+	if err := rewrite.TableDescs(tables, descriptorRewrites, overrideDBName); err != nil {
```
correct, if you're doing `BACKUP DATABASE xxx ..... WITH new_db_name = xxxx`, then db names in views are not rewritten. I'll add a test case for that actually, even it shares same code as udf.

chengxiong-ruan replied:
test case added.

chengxiong-ruan commented:
> I think we want to add a check for USAGE during backup here: https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/backupccl/backup_planning.go#L466
> 
Thanks for the review! that's a good catch on the privilege check, fixed with test cases :)

> Overall this looks reasonable to me. I suppose that UDFs may references tables that aren't being backed up, but that matches how we handle foreign keys and such currently.
The idea is that UDFs can only be backed up at db or cluster level. And we don' allow cross-db reference from a UDF, so that shouldn't be a concern.

stevendanna reviewed (approved):
Thanks!

chengxiong-ruan commented:
TFTR!
bors r+

chengxiong-ruan commented:
bors r+
